### ISSUE TITLE
CCDirector doesn't call onEnter on next scene if last scene was a transition

### ISSUE THREAD

**Issue by wildbunny** [cocos2d/cocos2d-x#8716]
void Director::setNextScene()
{
    **\* INITIALLY SET FROM CURRENT SCENE ***
    bool runningIsTransition = dynamic_cast<TransitionScene*>(_runningScene) != nullptr;
    ....

  **\* RUNNING SCENE CHANGED HERE ***
   _runningScene = _nextScene;
    _nextScene->retain();
    _nextScene = nullptr;

```
*** CHECK IS NOW FOR LAST SCENE ***
if ((! runningIsTransition) && _runningScene)
{
     *** ON ENTER NEVER CALLED ON NEW SCENE IF LAST SCENE WAS TRANSITION
     _runningScene->onEnter();
....
```

I've copied and annotated the location in the source where the problem lies. If the last scene was a transition, the next scene does not have it's onEnter() function called. The runningIsTransition variable needs to be refreshed before the check occurs to fix the problem.

Cheers, Paul.

**Comment by dumganhar** [comment]
@wildbunny, thanks for your feedback. I will check this issue.

**Comment by dumganhar** [comment]
Hi @wildbunny, could you please paste some codes for how to reproduce this issue? Thanks.

**Comment by dumganhar** [comment]
Hi @wildbunny , the `runningIsTransition` flag is used for identify whether current running scene is a transition scene. If it is, `Director::setNextScene` is invoked from

``` c++
void TransitionScene::setNewScene(float dt)
{    
    CC_UNUSED_PARAM(dt);

    this->unschedule(CC_SCHEDULE_SELECTOR(TransitionScene::setNewScene));

    // Before replacing, save the "send cleanup to scene"
    Director *director = Director::getInstance();
    _isSendCleanupToScene = director->isSendCleanupToScene();

    director->replaceScene(_inScene);

    // issue #267
    _outScene->setVisible(true);
}
```

In this way, the onEnter of next scene should be called before the transition started.
I don't know why it doesn't work for you. If you could tell us how you're using transition, it will be more helpful for us to find out the problem if it's a problem. :)

**Comment by wildbunny** [comment]
Hi there,

The problem is that if you are changing scenes inside the
onEnterTransitionDidFinish, or onEnter callbacks,
then Director::setNextScene() will not call those callbacks if the
runningIsTransition flag is set. So if the current scene is a transition,
and the next scene is not, those important callbacks never get called.

Cheers, Paul.

On 24 November 2014 at 10:04, James Chen notifications@github.com wrote:

> Hi @wildbunny https://github.com/wildbunny , the runningIsTransition
> flag is used for identify whether current running scene is a transition
> scene. If it is, Director::setNextScene is invoked from
> 
> void TransitionScene::setNewScene(float dt)
> {
>     CC_UNUSED_PARAM(dt);
> 
> ```
> this->unschedule(CC_SCHEDULE_SELECTOR(TransitionScene::setNewScene));
> 
> // Before replacing, save the "send cleanup to scene"
> Director *director = Director::getInstance();
> _isSendCleanupToScene = director->isSendCleanupToScene();
> 
> director->replaceScene(_inScene);
> 
> // issue #267
> _outScene->setVisible(true);
> ```
> 
> }
> 
> In this way, the onEnter of next scene should be called before the
> transition started.
> I don't know why it doesn't work for you. If you could tell us how you're
> using transition, it will be more helpful for us to find out the problem if
> it's a problem. :)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cocos2d/cocos2d-x/issues/8716#issuecomment-64172456.

## 

Paul Firth
http://www.wildbunny.co.uk/blog
http://twitter.com/paul_m_firth

**Comment by dumganhar** [comment]
Hey @wildbunny , thanks for your reply.
I will try to reproduce this issue in your way.
But for workaround, I think you could use a schedule in `onEnter` or `onEnterTransitionDidFinish`.
And in the schedule callback, do the `replaceScene` stuff.

**Comment by wildbunny** [comment]
Hi James,

Thanks for the advice - for my use case, I actually edited the
CCDirector::setNextScene() function like this:

 if (/*(! runningIsTransition) && */_runningScene)
    {
        _runningScene->onEnter();
        _runningScene->onEnterTransitionDidFinish();
    }

which has had no ill effects so far :)

On 24 November 2014 at 10:34, James Chen notifications@github.com wrote:

> Hey @wildbunny https://github.com/wildbunny , thanks for your reply.
> I will try to reproduce this issue in your way.
> But for workaround, I think you could use a schedule in onEnter or
> onEnterTransitionDidFinish.
> And in the schedule callback, do the replaceScene stuff.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cocos2d/cocos2d-x/issues/8716#issuecomment-64175532.

## 

Paul Firth
http://www.wildbunny.co.uk/blog
http://twitter.com/paul_m_firth

**Comment by dumganhar** [comment]
+@ricardoquesada , is it reasonable to do a `replaceScene` operation in `onEnter`, `onEnterTransitionDidFinish`, `onExit` or `onExitTransitionDidStart`?

**Comment by wildbunny** [comment]
@dumganhar How else are we expected to chain sequences of transitions together?

At the very least, the code should CCASERT that a replaceScene has not occurred in a callback in which it is forbidden?

**Comment by dumganhar** [comment]
@wildbunny , so you're chaining sequences of transitions by running another transition scene in 

``` c++
onEnterTransitionDidFinish
```

?
Yep, I think doing `replaceScene` in onEnterTransitionDidFinish is reasonable. :)
We need to investigate and find a better solution. :) Thanks.

**Comment by wildbunny** [comment]
Indeed! There appears to be no other clean way to chain transitions
together that I could find :)

On 25 November 2014 at 10:23, James Chen notifications@github.com wrote:

> @wildbunny https://github.com/wildbunny , so you're chaining sequences
> of transitions by running another transition scene in
> 
> onEnterTransitionDidFinish
> 
> ?
> Yep, I think doing replaceScene in onEnterTransitionDidFinish is
> reasonable. :)
> We need to investigate and find a better solution. :) Thanks.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cocos2d/cocos2d-x/issues/8716#issuecomment-64376348.

## 

Paul Firth
http://www.wildbunny.co.uk/blog
http://twitter.com/paul_m_firth
